st2common/st2common/models/utils/action_param_utils.py

```diff
@@ -70,3 +79,51 @@ def is_immutable(param_meta):
     immutable_params = {k: merged_params[k] for k in immutable}
 
     return (required_params, optional_params, immutable_params)
+
+
+def cast_params(action_ref, params):
+    """
+    """
+    def cast_object(x):
+        if isinstance(x, str) or isinstance(x, unicode):
+            try:
+                return json.loads(x)
+            except:
+                return ast.literal_eval(x)
+        else:
+            return x
+
+    casts = {
+        'array': (lambda x: ast.literal_eval(x) if isinstance(x, str) or isinstance(x, unicode)
+                  else x),
+        'boolean': (lambda x: ast.literal_eval(x.capitalize())
+                    if isinstance(x, str) or isinstance(x, unicode) else x),
+        'integer': int,
+        'number': float,
+        'object': cast_object,
+        'string': str
+    }
+    action_db = action_db_util.get_action_by_ref(action_ref)
+    action_parameters_schema = action_db.parameters
+    runnertype_db = action_db_util.get_runnertype_by_name(action_db.runner_type['name'])
+    runner_parameters_schema = runnertype_db.runner_parameters
+    # combine into 1 list of parameter schemas
+    parameters_schema = {}
+    if runner_parameters_schema:
+        parameters_schema.update(runner_parameters_schema)
+    if action_parameters_schema:
+        parameters_schema.update(action_parameters_schema)
+    # cast each param individually
+    for k, v in six.iteritems(params):
+        parameter_schema = parameters_schema.get(k, None)
+        if not parameter_schema:
+            continue
+        parameter_type = parameter_schema.get('type', None)
+        if not parameter_type:
+            continue
+        cast = casts.get(parameter_type, None)
+        LOG.debug('Casting param: %s of type %s to type: %s', v, type(v), parameter_type)
+        if not cast:
+            continue
```

Comment:
The types we support are per json schema. I don't believe unicode is present in json-schema in the types.

However, if you are talking about how casting unicode to str if the string value is in unicode then yeah that is missed.